ext/priority_sampling/priority_sampling.c

```diff
@@ -63,55 +61,159 @@ static bool dd_rule_matches(zval *pattern, zval *prop) {
+                return false;
+            }
+        }
+    }
+    if ((rule_pattern = zend_hash_str_find(rule, ZEND_STRL("name")))) {
+        if (!dd_rule_matches(rule_pattern, &span->property_name)) {
+            return false;
+        }
+    }
+    if ((rule_pattern = zend_hash_str_find(rule, ZEND_STRL("resource")))) {
+        if (!dd_rule_matches(rule_pattern, &span->property_resource)) {
+            return false;
+        }
+    }
+    if ((rule_pattern = zend_hash_str_find(rule, ZEND_STRL("tags"))) && Z_TYPE_P(rule_pattern) == IS_ARRAY) {
+        zend_array *tag_rules = Z_ARR_P(rule_pattern);
+        zend_array *meta = ddtrace_property_array(&span->property_meta);
+        zend_string *tag_name;
+        ZEND_HASH_FOREACH_STR_KEY_VAL(tag_rules, tag_name, rule_pattern) {
+            if (tag_name) {
+                zval *value;
+                if (!(value = zend_hash_find(meta, tag_name))) {
+                    return false;
+                }
+                if (!dd_rule_matches(rule_pattern, value)) {
+                    return false;
+                }
+            }
+        } ZEND_HASH_FOREACH_END();
+    }
+
+    return true;
+}
+
+// If there is one rule matching in *ANY* span, then all further rules are ignored.
+// Thus we check only rules until a specific index then.
+static ddtrace_rule_result dd_match_rules(ddtrace_span_data *span, bool eval_root, int skip_at) {
+    int index = -3;
 
-        zval *service = &span->property_service;
+    if (++index >= skip_at) {
+        return (ddtrace_rule_result){ .sampling_rate = 0, .rule = INT32_MAX };
+    }
+
+    zend_array *meta = ddtrace_property_array(&span->property_meta);
+    if (zend_hash_str_exists(meta, ZEND_STRL("manual.keep"))) {
+        return (ddtrace_rule_result){ .sampling_rate = 1, .rule = -2 };
+    }
+
+    if (++index >= skip_at) {
+        return (ddtrace_rule_result){ .sampling_rate = 0, .rule = INT32_MAX };
+    }
+    if (zend_hash_str_exists(meta, ZEND_STRL("manual.drop"))) {
+        return (ddtrace_rule_result){ .sampling_rate = 0, .rule = -1 };
+    }
+
+    zval *rule;
+    ZEND_HASH_FOREACH_VAL(get_DD_TRACE_SAMPLING_RULES(), rule) {
+        if (++index >= skip_at) {
+            break;
+        }
+
+        if (Z_TYPE_P(rule) != IS_ARRAY) {
+            continue;
+        }
 
-        ZEND_HASH_FOREACH_VAL(get_DD_TRACE_SAMPLING_RULES(), rule) {
-            if (Z_TYPE_P(rule) != IS_ARRAY) {
+        if (!eval_root) {
+            zval *applies = zend_hash_str_find(Z_ARR_P(rule), ZEND_STRL("target_span"));
+            if (!applies || Z_TYPE_P(applies) != IS_STRING || !zend_string_equals_literal(Z_STR_P(applies), "any")) {
                 continue;
             }
+        }
 
-            bool rule_matches = true;
+        zval *sample_rate_zv;
+        if (dd_check_sampling_rule(Z_ARR_P(rule), span) && (sample_rate_zv = zend_hash_str_find(Z_ARR_P(rule), ZEND_STRL("sample_rate")))) {
```

Comment:
Hum, I'm not sure that it is taking into account that the default value of `sample_rate` is `1.0`. From what I see/understand, currently, if a rule doesn't include the `sample_rate` tag, then even if `dd_check_sampling_rule` evaluates to true, the sample rate won't be changed. I believe case 2 or 4A from the RFC would fail.

Reply:
I see, it defaults to 1, right.

Reply:
We already required sample_rate to be present in the past, and it's not been an explicit change in the RFC, so it must have been always wrong :-)